TCF4.py

```diff
@@ -0,0 +1,121 @@
+from Bio import SeqIO
+from Bio.SeqUtils import GC
+
+def filter_fastq(input_path: str, quality_threshold: int, output_filename="final_filtered.fastq",  gc_bounds=(40, 60), length_bounds=(50, 350)):
+    filename = input_path
+    records = SeqIO.parse(filename, "fastq")
+    ###quality filter
+    good_reads = (rec for rec in records if min(rec.letter_annotations["phred_quality"]) >= quality_threshold)
+    result_quality = SeqIO.write(good_reads, "good_quality.fastq", "fastq")
+    result_quality_GC = SeqIO.parse("good_quality.fastq", "fastq")
+    ###GC content filter
+    min_gc_content = gc_bounds[0]
+    max_gc_content = gc_bounds[1]
+    GC_quality_filt = []
+    
+    for sequence in result_quality_GC:
+        if min_gc_content <= GC(sequence.seq) <= max_gc_content:
+            GC_quality_filt.append(sequence)
+            
+    result_quality = SeqIO.write(GC_quality_filt, "good_quality_GC.fastq", "fastq")
+    result_quality_GC_length = SeqIO.parse("good_quality_GC.fastq", "fastq")
+    
+    ##length filter
+    filtered_GC_quality_length = []
+    
+    for sequence in result_quality_GC_length:
+        if len(sequence.seq) >= length_bounds[0] and len(sequence.seq) <= length_bounds[1]:
+            filtered_GC_quality_length.append(sequence)
+            
+    result_quality = SeqIO.write(filtered_GC_quality_length, output_filename, "fastq")
+    
+    print(result_quality)
+
+#filter_fastq("example_fastq.fastq", 15)
+
+
+from abc import ABC, abstractmethod
+
+class InvalidInputError(ValueError):
+    pass
+
+class BiologicalSequence(ABC, str):
+    @abstractmethod
+    def __init__(self, seq):
+        self.seq = seq
+        
+    def __len__(self):
+        return len(self.seq)
+    
+    def __getitem__(self, index):
+        return self.seq[int(index)]
+    
+    def __repr__(self):
+        return __str__(self.seq)
+    
+    def check_nucleic_acid(self):
+        unique_chars = set(self.seq)
+        nucleotides_dna = set('ATGCatgc')
+        nucleotides_rna = set('AUGCaugc')
+        if unique_chars <= nucleotides_dna:
+            seq = 'dna'
+        elif unique_chars <= nucleotides_rna:
+            seq = 'rna'
+        else:
+            raise InvalidInputError()
+            return seq_type
```

Comment:
Идея правильная - метод сравнивает уникальные элементы с "алфавитом" конкретного полимера. Но:

1. `unique_chars = set(self.seq)` предполагает, что `self.seq` у тебя итерируемый, а ты этого явно нигде не проверяешь. Т.е. если создать экземпляр, передав `DNASequence` число `1000`, эта проверка упадёт не по предусмотренному тобой сценарию.
2. получается, ты внутри абстрактного класса реализуешь проверку только для двух конкретных наследников. Хотя, например, такая логика применима и к аминокислотным, но для них этот метод работать не будет. Логично реализовать проверку, запрашивая у класса его алфавит. При этом, в `complement` у тебя всё отлично, и метод опирается на атрибуты класса, для которого его вызывают.  
Здесь можно сделать похожим образом - добавить классу атрибут `self.alphabet` и сделать как-то так:
```suggestion
    def check_alphabet(self) -> bool:
        return set(self.seq).issubset(self.alphabet)
```
Теперь все потомки `BiologicalSequence` будут при проверке сравнивать свой собственный алфавит и свою собственную последовательность.